feat(eslint-plugin-template): [button-has-type] add rule

chernodub opened the pull request:
closes #823

chernodub commented:
Hi @JamesHenry, I'd appreciate if you could look into this PR. So that we could continue working on it or, at least, close it. Thanks!

JamesHenry commented on packages/eslint-plugin-template/tests/rules/button-has-type/cases.ts:
```diff
@@ -0,0 +1,71 @@
+const missingType: MessageIds = 'missingType';
+const invalidType: MessageIds = 'invalidType';
+
+const validCasesWithTextAttribute = VALID_BUTTON_TYPES.map(
```
Please remove these abstractions in favour of duplication and keep everything inline in the valid cases array. Folks shouldn't have to parse test cases as they read or update them, it adds unnecessary mental overhead and harms the readability of the tests

JamesHenry commented on packages/eslint-plugin-template/src/rules/button-has-type.ts:
```diff
@@ -0,0 +1,126 @@
+  );
+}
+
+function getInvalidButtonTypeIfPresent(
```
I found this difficult to read with there being nested closures which do not leverage any aspect of their outer scope. Additionally `lookIn` is not a well established semantic prefix (unlike `get` or `find` for example) so it's way more open to personal interpretation - I wouldn't have necessarily expected a function starting with `lookIn` to return a value, for example...

Speaking of returning a value, you are directly returning a `TextAttribute` value to represent a type of `InvalidButtonTypeInfo` in `lookInTextAttributes` instead of creating its more specific shape (unlike what you do in the `lookInBoundAttributes` function just below).

In summary then I think keeping it simple would be preferable here, e.g. something like:

```ts
function getInvalidButtonTypeIfPresent(
  element: TmplAstElement,
): InvalidButtonTypeInfo | null {
  const invalidTextAttribute = element.attributes.find(
    ({ name, value }) =>
      name === TYPE_ATTRIBUTE_NAME && !VALID_BUTTON_TYPES.includes(value),
  )
  if (invalidTextAttribute) {
    return {
      value: invalidTextAttribute.value,
      sourceSpan: invalidTextAttribute.sourceSpan,
    };
  }

  for (const { name, value, sourceSpan } of element.inputs) {
    // Intentionally ignore the property binding by looking for literal primitives only
    // in order to avoid type-checking for the property, e.g. check that it's 'submit', 'button', etc.
    if (
      name === TYPE_ATTRIBUTE_NAME &&
      value instanceof ASTWithSource &&
      value.ast instanceof LiteralPrimitive &&
      !VALID_BUTTON_TYPES.includes(value.ast.value)
    ) {
      return {
        value: value.ast.value,
        sourceSpan,
      };
    }
  }

  return null;
}
```

JamesHenry commented on packages/eslint-plugin-template/src/rules/button-has-type.ts:
```diff
@@ -0,0 +1,126 @@
+export const INVALID_TYPE_DATA_KEY = 'type';
+export type MessageIds = 'invalidType' | 'missingType';
+
+interface InvalidButtonTypeInfo {
```
I really don't think these comments add any value here (especially the source span) and they can be removed

JamesHenry reviewed (changes requested):
I'm very sorry for the delay in getting back to you on this @chernodub and thank you so much for taking the initiative and contributing this rule.

Please take a look at the review comments. You will also need to run `yarn update-rule-docs` in order to generate the necessary rule documentation to accompany this rule.

chernodub commented:
Thanks a lot for a valuable feedback @JamesHenry! 🙏

I've updated the code following your comments (agreed to all the points you mentioned). I've also rebased the changes on `origin/master` branch, so it's ready for merging.

JamesHenry commented:
Thanks a lot @chernodub! I'll release this on v13 so you're not forced to upgrade to v14 of Angular to use it
